### ISSUE TITLE
Use more convenient dates and times

### ISSUE THREAD

**Issue by bcudini** [geier/khal#169]

**Comment by geier** [comment]
This is a good idea and I've thought about this myself.

The problem with using GNU date is, that while it my be legal to import GPL licensed code I'm not sure its not against the wishes of the author. So I would prefer a MIT/BSD etc. licensed package to be used. An apache licensed module for which could be helpful is [parsedatetime](https://github.com/bear/parsedatetime).

For the where, have a look at `generate_random_uid()` in `khal/aux.py`.

If you have any more questions, don't hesitate to ask. I'd love to see a pull request for this.

**Comment by untitaker** [comment]
I'd consider using GNU tools a valid option re legal situation, but doing so might harm portability.

**Comment by untitaker** [comment]
You might also want to do this the way GNOME California does: Just take a description, and extract the dt from there if found. We should also add CLI options for enforcing the date with strict behavior (for scripts invoking khal new), and make the user use that if we fail to extract a date.

**Comment by geier** [comment]
[California's quick add](https://wiki.gnome.org/Apps/California/HowToUseQuickAdd) looks pretty nice, lots of good ideas waiting to be stolen

**Comment by bcudini** [comment]
The parsedatetime module looks good, california seems even better but their code is unusable here (looks like it's written in Vala and GPL licenced).

For the where, did you mean `construct_event()` in `khal/aux.py`, the function just after `generate_random_uid()`? I'm looking into it right now and will try to have a working PoC asap

**Comment by geier** [comment]
@bcudini yes, I did mean `construct event()`, I don't know how that error happened (corrected it now)

**Comment by geier** [comment]
I just had a look at california's [details parser](https://git.gnome.org/browse/california/tree/src/component/component-details-parser.vala). While it is obviously a lot better than what we currently have, their approach is at least similar: *Tokenize* the input string and then use different parsers for different components (start, end, duration, location etc) on that stack, popping elements if they are 'used'.

**Comment by geier** [comment]
If you want to help with this issue, just write down all variants you can think of a clever quick add function should understand.

**Comment by hobarrera** [comment]
AFAIK, California's implementation was rather broken in that it expected US-formated date (eg: "Monday 6pm") and did not work with standard time formats ("Monday 1800hs", "Monday 1800", etc), as well as some other really common variations.. However, the high level idea should work.

> If you want to help with this issue, just write down all variants you can think of a clever quick add function should understand.

Possible formats for input:

 * Beer with Alice tomorrow 1400
 * Lunch with bob in two days midday
 * Breakfast with Alice Friday quarter to twelve
 * Party next Wednesday at midnight
 * Beer with Peter on March 2nd
 * Meeting in two hours until midnight
 * Party tonight at 2000 (note that tonight equals today)
 * Beer with Alice in one week

Recurring events sound a bit trickier, but are basically determined by the way `every <X> [other] <Y>`, `X` being an integer, and `Y` can be month, day, week, etc. There's also:

 * Beer with Alice every other day starting tomorrow
 * Beer with Bob every week
 * Beer with Charlie every 2 months.

Non-infinite recurring events:

 * Lunch with Mum every Thrusday until June 1st
 * Call Mum every day until the end of the year
 * Beer with Alice every day from 20 to 22 until next Saturday

Ironically, I don't drink as much beer as you'd expect, they were just the quickest examples. 😄

**Comment by geier** [comment]
To anyone who wants to start working on this, I would recommend having a look at [spacy](http://spacy.io/) (which [has just been relicensed under MIT license](http://spacy.io/blog/spacy-now-mit/)), to me it looks like the most practical solution for natural language processing (I only did some quick experiments with it though).

**Comment by geier** [comment]
another interesting looking library (haven't tried it yet): https://github.com/akoumjian/datefinder

**Comment by hobarrera** [comment]
`spacy` looks great, but not really suited to our needs, regrettably:

> `__init__`: Load the linguistic analysis pipeline. Loading may take up to a minute, and the instance consumes 2 to 3 gigabytes of memory.

[[ref](https://spacy.io/docs)]

**Comment by geier** [comment]
@hobarrera alright, that doesn't look like it fits our use case.

**Comment by untitaker** [comment]
`parsedatetime` works very well in todoman for me.
